File: Project3_Python/Animals/Turtle.py
from Animal import Animal

import random
from Organism import OrganismsEnum


class Turtle(Animal):
    __DEFAULT_INITIATIVE = 1
    __DEFAULT_STRENGTH = 2
    __TURTLE_DEFENSE = 5
    __THE_FLOOR_OF_TURTLE_MOVEMENT = 75

    def __init__(self, world, position, strength=__DEFAULT_STRENGTH):
        super().__init__(world, position, strength)
        self.type_of_organism = OrganismsEnum.Turtle
        self._initiative = self.__DEFAULT_INITIATIVE

    def collision(self, attacker_organism):
        probability_of_escape = random.randint(0, 99)
        attacker_organism.put_into_world()

        if self.type_of_organism == attacker_organism.get_type_of_organism():
            self.multiply_organism()
            return

        if probability_of_escape < self.THE_FLOOR_OF_ESCAPE_ANTELOPE:
            new_position = self.get_new_position()
            if new_position.x != -1 and self._world.check_if_not_occupied(new_position):
                if attacker_organism.get_type_of_organism() == OrganismsEnum.Antelope:
                    attacker_organism.change_position(new_position)
                    attacker_organism.put_into_world()
                    self.put_into_world()
                    self._world.add_comment(str(attacker_organism) + " escaped.")
                    return

        if self.__TURTLE_DEFENSE > attacker_organism.get_strength():
            attacker_organism.exchange_positions()
            attacker_organism.put_into_world()
            self.put_into_world()
            self._world.add_comment(str(self) + " deflected the attack.")

        if self.strength > attacker_organism.get_strength():
            self.put_into_world()
            if attacker_organism.get_type_of_organism() != OrganismsEnum.Human:
                self._world.remove_organism_from_list(attacker_organism)
            else:
                self._world.delete_human()
            self._world.add_comment(str(self) + " won with " + str(attacker_organism))
        else:
            if self.type_of_organism != OrganismsEnum.Human:
                self._world.remove_organism_from_list(self)
            else:
                self._world.delete_human()
            self._world.add_comment(str(attacker_organism) + " won with " + str(self))

    def action(self):
        new_position = self.get_new_position()
        possibility_of_movement = random.randint(0, 99)
        if new_position.x == -1 or possibility_of_movement < self.__THE_FLOOR_OF_TURTLE_MOVEMENT:
            return False
        self._prev_position = self._position
        self._position = new_position
        self.try_to_go_to_position()
        return True

    def __str__(self):
        return "Turtle"

File: Project2_Java/src/pl/pg/oop/project2/organisms/animals/Turtle.java
package pl.pg.oop.project2.organisms.animals;

import pl.pg.oop.project2.organisms.Animal;
import pl.pg.oop.project2.organisms.Organism;
import pl.pg.oop.project2.organisms.OrganismsEnum;
import pl.pg.oop.project2.organisms.World;

import java.awt.*;
import java.util.Random;

public class Turtle extends Animal {
    private final static int DEFAULT_INITIATIVE = 1;
    private final static int DEFAULT_STRENGTH = 2;
    private final static int TURTLE_DEFENSE = 5;
    private final static int THE_FLOOR_OF_TURTLE_MOVEMENT = 75;
    public Turtle(World world, Point position) {
        super(world, position);
        this.TypeOfOrganism = OrganismsEnum.TURTLE;
        this.initiative = DEFAULT_INITIATIVE;
        this.strength = DEFAULT_STRENGTH;
    }

    public Turtle(World world, Point position, int strength) {
        super(world, position, strength);
        this.TypeOfOrganism = OrganismsEnum.TURTLE;
        this.initiative = DEFAULT_INITIATIVE;
    }

    @Override
    public void collision(Organism AttackerOrganism) {
        Random random = new Random();
        int ProbabilityOfEscape = random.nextInt(100);
        AttackerOrganism.PutIntoWorld();
        if(TypeOfOrganism == AttackerOrganism.getTypeOfOrganism()) {
            MultiplyOrganism();
            return;
        }
        if(ProbabilityOfEscape < THE_FLOOR_OF_ESCAPE_ANTELOPE) {
            Point newPosition = getNewPosition();
            if(newPosition.x != -1 && world.checkIfNotOccupied(newPosition)) {
                if (AttackerOrganism.getTypeOfOrganism() == OrganismsEnum.ANTELOPE) {
                    AttackerOrganism.ChangePosition(newPosition);
                    AttackerOrganism.PutIntoWorld();
                    PutIntoWorld();
                    world.addComment(AttackerOrganism + " escaped.");
                    return;
                }
            }
        }
        if(TURTLE_DEFENSE > AttackerOrganism.getStrength()) {
            AttackerOrganism.exchangePositions();
            AttackerOrganism.PutIntoWorld();
            PutIntoWorld();
            world.addComment(this + " deflected the attack.");
        }
        if(strength > AttackerOrganism.getStrength()) {
            PutIntoWorld();
            if(AttackerOrganism.getTypeOfOrganism() != OrganismsEnum.HUMAN)
                world.RemoveOrganismFromList(AttackerOrganism);
            else
                world.DeleteHuman();
            world.addComment(this + " won with " + AttackerOrganism);
        }
        else {
            if(this.TypeOfOrganism != OrganismsEnum.HUMAN)
                world.RemoveOrganismFromList(this);
            else
                world.DeleteHuman();
            world.addComment(AttackerOrganism + " won with " + this);
        }
    }

    @Override
    public boolean action() {
        Point newPosition = getNewPosition();
        Random random = new Random();
        int PossibilityOfMovement = random.nextInt(100);
        if(newPosition.x == -1 || PossibilityOfMovement < THE_FLOOR_OF_TURTLE_MOVEMENT) return false;
        prevPosition = position;
        position = newPosition;
        TryToGoToPosition();
        return true;
    }

    @Override
    public String toString() {
        return "Turtle";
    }
}
